File: backend/app/schemas/growth.py
"""Historical growth analytics API schemas."""

from datetime import date

from pydantic import BaseModel, Field


class CreatorGrowthItem(BaseModel):
    creator_name: str
    youtube_channel_id: str = ""
    subscribers_now: int = 0
    subscribers_delta_7d: int = 0
    subscribers_delta_30d: int = 0
    growth_7d_pct: float = 0.0
    growth_30d_pct: float = 0.0
    total_views_now: int = 0
    views_delta_7d: int = 0
    velocity_views_per_day: float = 0.0
    snapshot_days: int = 0


class CreatorGrowthResponse(BaseModel):
    items: list[CreatorGrowthItem] = Field(default_factory=list)
    snapshot_date_latest: date | None = None


class VideoBreakoutItem(BaseModel):
    video_id: int
    title: str = ""
    creator_name: str = ""
    views_now: int = 0
    views_delta_7d: int = 0
    growth_7d_pct: float = 0.0
    velocity_views_per_day: float = 0.0
    breakout_score: float = 0.0


class VideoBreakoutsResponse(BaseModel):
    items: list[VideoBreakoutItem] = Field(default_factory=list)


class VelocityItem(BaseModel):
    video_id: int
    title: str = ""
    creator_name: str = ""
    views_now: int = 0
    velocity_views_per_day: float = 0.0
    views_delta_7d: int = 0


class VelocityResponse(BaseModel):
    items: list[VelocityItem] = Field(default_factory=list)

File: frontend/types/growth.ts
/** Historical growth analytics — mirrors backend /analytics growth endpoints */

export interface CreatorGrowthItem {
  creator_name: string;
  youtube_channel_id: string;
  subscribers_now: number;
  subscribers_delta_7d: number;
  subscribers_delta_30d: number;
  growth_7d_pct: number;
  growth_30d_pct: number;
  total_views_now: number;
  views_delta_7d: number;
  velocity_views_per_day: number;
  snapshot_days: number;
}

export interface CreatorGrowthResponse {
  items: CreatorGrowthItem[];
  snapshot_date_latest: string | null;
}

export interface VideoBreakoutItem {
  video_id: number;
  title: string;
  creator_name: string;
  views_now: number;
  views_delta_7d: number;
  growth_7d_pct: number;
  velocity_views_per_day: number;
  breakout_score: number;
}

export interface VideoBreakoutsResponse {
  items: VideoBreakoutItem[];
}

export interface VelocityItem {
  video_id: number;
  title: string;
  creator_name: string;
  views_now: number;
  velocity_views_per_day: number;
  views_delta_7d: number;
}

export interface VelocityResponse {
  items: VelocityItem[];
}
